Fixing Python 3.6 deprecations

MSeifert04 opened the pull request:
Like #5484 but instead of ignoring deprecation warnings this PR fixes those.

MSeifert04 commented:
Looks like pytest 2.8.3 still uses `raises StopIteration` which isn't supported in py36 anymore. The other failures are identical to #5484

mhvk commented on astropy/modeling/functional_models.py:
```diff
@@ -288,7 +288,7 @@ class Gaussian2D(Fittable2DModel):
             \\sigma_x^2               & \\rho \\sigma_x \\sigma_y \\\\
             \\rho \\sigma_x \\sigma_y & \\sigma_y^2
-            \end{array}\\right)
+            \\end{array}\\right)
```
Ideally, one would change this one to a raw string too...

MSeifert04 replied:
This docstring contains a lot of formulas but only one error so I've just fixed the error.

Will change it later.

mhvk commented:
Great clean-up!

MSeifert04 commented on astropy/utils/console.py:
```diff
@@ -365,7 +365,7 @@ def strip_ansi_codes(s):
     Remove ANSI color codes from the string.
     """
-    return re.sub('\033\[([0-9]+)(;[0-9]+)*m', '', s)
+    return re.sub(r'\033\[([0-9]+)(;[0-9]+)*m', '', s)
```
I'm not sure about this change: will the leading `\033` interpreted correctly as raw string?

mhvk replied:
No, surely not:
```
In [11]: '\033\[([0-9]+)(;[0-9]+)*m',
Out[11]: ('\x1b\\[([0-9]+)(;[0-9]+)*m',)

In [12]: r'\033\[([0-9]+)(;[0-9]+)*m',
Out[12]: ('\\033\\[([0-9]+)(;[0-9]+)*m',)
```

MSeifert04 replied:
ok, will change this to a interpreted string with escaped `\\` again. Thanks for testing 👍

MSeifert04 replied:
hm, I'm actually not sure if it's a big deal in regular expressions:

```
>>> import re

>>> re.search("\003", "mister\003smith")
<_sre.SRE_Match object; span=(6, 7), match='\x03'>

>>> re.search(r"\003", "mister\003smith")
<_sre.SRE_Match object; span=(6, 7), match='\x03'>
```

mhvk replied:
Is this in the new python?
Anyway, might as well try to have it make most sense. Certainly:
```
re.search("\003", r"mister\003smith")
```
does not return any matches

It does seem a bit troublesome that
```
re.search(r"\003", r"mister\003smith")
```
also does not return any matches... While
```
re.search(r"\\003", r"mister\003smith")
Out[7]: <_sre.SRE_Match object; span=(6, 10), match='\\003'>
```

MSeifert04 replied:
I don't know if I should laugh or cry ... 😄

mhvk replied:
The lesson may be that it does *not* pay to "make things easier for the user" -- it just ends being confusing.

MSeifert04 commented:
ok, all Errors and Deprecations are fixed.

There are still 6 errors which I don't know how to fix. All of them somewhat related to catched Warnings, but I can't find any note that they changed something regarding Warnings in 3.6.

```
=================================== FAILURES ===================================

____________________________ test_errwarn_reporting ____________________________

recwarn = <_pytest.recwarn.WarningsRecorder object at 0x7f1a8ac367f0>

    def test_errwarn_reporting(recwarn):

        """

        Test that the ERFA error reporting mechanism works as it should

        """

    

        # no warning

        erfa.dat(1990, 1, 1, 0.5)

    

        # check warning is raised for a scalar

        erfa.dat(100, 1, 1, 0.5)

>       w = recwarn.pop(erfa.ErfaWarning)

E       AssertionError: <class 'astropy._erfa.core.ErfaWarning'> not found in warning list

astropy/_erfa/tests/test_erfa.py:106: AssertionError

_________________ TestKernels.test_kernel_normalization_large __________________

self = <astropy.convolution.tests.test_kernel_class.TestKernels object at 0x7f1a8a0bc908>

recwarn = <_pytest.recwarn.WarningsRecorder object at 0x7f1a8a0bc9e8>

    def test_kernel_normalization_large(self, recwarn):

        """

            Test that a warning is issued when the inverse normalization factor

            is large.

            """

        kernel = CustomKernel(np.ones(3) * 1.e-3)

        kernel.normalize()

>       w = recwarn.pop(AstropyUserWarning)

E       AssertionError: <class 'astropy.utils.exceptions.AstropyUserWarning'> not found in warning list

astropy/convolution/tests/test_kernel_class.py:483: AssertionError

______________________________ test_future_altaz _______________________________

recwarn = <_pytest.recwarn.WarningsRecorder object at 0x7f1a8e64b860>

    def test_future_altaz(recwarn):

        """

        While this does test the full stack, it is mostly meant to check that a

        warning is raised when attempting to get to AltAz in the future (beyond

        IERS tables)

        """

        from ...utils.exceptions import AstropyWarning

    

        # this is an ugly hack to get the warning to show up even if it has already

        # appeared

        from ..builtin_frames import utils

        if hasattr(utils, '__warningregistry__'):

            utils.__warningregistry__.clear()

    

        location = EarthLocation(lat=0*u.deg, lon=0*u.deg)

        t = Time('J2161')

    

        SkyCoord(1*u.deg, 2*u.deg).transform_to(AltAz(location=location, obstime=t))

    

        # check that these message(s) appear among any other warnings.  If tests are run with

        # --remote-data then the IERS table will be an instance of IERS_Auto which is

        # assured of being "fresh".  In this case getting times outside the range of the

        # table does not raise an exception.  Only if using IERS_B (which happens without

        # --remote-data, i.e. for all CI testing) do we expect another warning.

        messages_to_find = ["Tried to get polar motions for times after IERS data is valid."]

        if isinstance(iers.IERS_Auto.iers_table, iers.IERS_B):

            messages_to_find.append("(some) times are outside of range covered by IERS table.")

    

        messages_found = [False for _ in messages_to_find]

        for w in recwarn.list:

            if issubclass(w.category, AstropyWarning):

                for i, message_to_find in enumerate(messages_to_find):

                    if message_to_find in str(w.message):

                        messages_found[i] = True

>       assert all(messages_found)

E       assert all([False])

astropy/coordinates/tests/test_iau_fullstack.py:175: AssertionError

____________________________ test_warnings_logging _____________________________

    def test_warnings_logging():

    

        # Without warnings logging

        with catch_warnings() as warn_list:

            with log.log_to_list() as log_list:

                warnings.warn("This is a warning", AstropyUserWarning)

        assert len(log_list) == 0

        assert len(warn_list) == 1

        assert warn_list[0].message.args[0] == "This is a warning"

    

        # With warnings logging

        with catch_warnings() as warn_list:

            log.enable_warnings_logging()

            with log.log_to_list() as log_list:

                warnings.warn("This is a warning", AstropyUserWarning)

            log.disable_warnings_logging()

>       assert len(log_list) == 1

E       assert 0 == 1

E        +  where 0 = len([])

astropy/tests/test_logger.py:87: AssertionError

___________________ test_warnings_logging_with_custom_class ____________________

    def test_warnings_logging_with_custom_class():

        class CustomAstropyWarningClass(AstropyWarning):

            pass

    

        # With warnings logging

        with catch_warnings() as warn_list:

            log.enable_warnings_logging()

            with log.log_to_list() as log_list:

                warnings.warn("This is a warning", CustomAstropyWarningClass)

            log.disable_warnings_logging()

>       assert len(log_list) == 1

E       assert 0 == 1

E        +  where 0 = len([])

astropy/tests/test_logger.py:126: AssertionError

_________________ test_warning_logging_with_io_votable_warning _________________

    def test_warning_logging_with_io_votable_warning():

        from ..io.votable.exceptions import W02, vo_warn

    

        with catch_warnings() as warn_list:

            log.enable_warnings_logging()

            with log.log_to_list() as log_list:

                vo_warn(W02, ('a', 'b'))

            log.disable_warnings_logging()

>       assert len(log_list) == 1

E       assert 0 == 1

E        +  where 0 = len([])

astropy/tests/test_logger.py:141: AssertionError

= 6 failed, 8246 passed, 399 skipped, 71 xfailed, 1 xpassed in 330.16 seconds ==

The command "python setup.py test --remote-data" exited with 1.

Done. Your build exited with 1.
```

mhvk commented:
@MSeifert04 - this is weird, since `catch_warnings` is used all over the place; maybe it is the interaction with the logger...

Now, `recwarn` is used only four times, and three give a problem.  The fourth is in  `astropy/coordinates/tests/test_regression.py`, where there is an [ugly hack](https://github.com/astropy/astropy/blob/master/astropy/coordinates/tests/test_regression.py#L195) to make the warning appear. Possibly that is the issue in the other cases as well, that a warning had already appeared earlier? Anyway, maybe it is best to just use `catch_warnings` everywhere?

mhvk commented:
For the logger, cc @astrofrog, @eteq, @adamginsburg (all of whom are "blamed" by git).

mhvk commented on astropy/wcs/wcs.py:
```diff
@@ -1898,7 +1898,7 @@ def all_world2pix(self, *args, **kwargs):
                distortions of all HST instruments, testing has
                shown that enabling this option would lead to a about
-               50-100\% penalty in computational time (depending on
+               50-100\\% penalty in computational time (depending on
```
why does this need a double `\\`? I thought the point here was to include just a `%` sign (which is not used for formatting). [note: am not sure any backslash is necessary at all.]

MSeifert04 replied:
`\%` is deprecated (the only allowed escaped sequences are listed in this [table](https://docs.python.org/2/reference/lexical_analysis.html#string-literals)) so it throws an Exception. That's why I escaped the first `\`.

Just remove (both) `\`?

mhvk commented on astropy/convolution/tests/test_kernel_class.py:
```diff
@@ -473,15 +473,16 @@ def test_kernel_normalization_mode(self):
-        kernel.normalize()
-        w = recwarn.pop(AstropyUserWarning)
-        assert issubclass(w.category, AstropyWarning)
+        with pytest.warns(AstropyUserWarning) as w:
```
I won't argue with code that works, though since everywhere else we seem to use `catch_warnings`, maybe that would be best here too? (Or possibly, we should deprecate that if `pytest` can handle it anyway -- but that would be for another PR, obviously.)

MSeifert04 replied:
It didn't even work - now I'm using `catch_warnings`.

MSeifert04 commented:
@mhvk Really nice suggestion to use `catch_warnings`. 👍 That eliminated 3 of the 6 failures.

mhvk commented:
I tried looking at the logger code, but cannot find what might go wrong... I think it might need some command-line debugging to see what is actually happening, and why the warning is not put on the log.

mhvk commented:
@MSeifert04 - I'd probably replace the last instance of `recwarn` with `catch_warnings` as well -- might as well make the code a little more internally consistent!

MSeifert04 commented:
> I'd probably replace the last instance of recwarn with catch_warnings as well -- might as well make the code a little more internally consistent!

ok, I removed those as well.

MSeifert04 commented:
No idea if that's because of a change in Pythons warning system, there were at least some changes:

- [warnings.py](https://github.com/python/cpython/commits/master/Lib/warnings.py)
- [warnings.c](https://github.com/python/cpython/commits/master/Python/_warnings.c)

 😓 

Maybe that's a better fit for the python issue tracker?

MSeifert04 commented:
That could also explain why the `recwarn` tests started to fail with 3.6

MSeifert04 commented:
@astrofrog @eteq You probably know the astropy-logger in more detail. If you have a hunch or suggestion feel free to add commits to this branch and/or post them here. 

I'm totally stuck here.

keflavich commented:
:+1: for the changes I looked at; it seems that all of the code I'm blamed for is just getting corrected string escapes.

MSeifert04 commented:
Sorry to ping again @astrofrog @eteq

I'm totally stuck here and if it's a bug with Python (I'm not saying it is!) we probably should mention it on their issue tracker so they have a fair change to fix it before the release (mid december if I remember correctly).

mhvk commented:
@MSeifert04 - should we split off the logging errors to a new issue (or, alternatively, split off the rest from this PR). I'd be quite happy to merge the other things that are improvements on their own accord.

MSeifert04 commented:
I'll rerun the original tests and the tests with 3.6. I would prefer to wait for the first beta release (9.12.) so they don't revert any changes or deprecations.

mhvk commented:
OK, fine to wait for the first beta.

astrofrog commented:
@MSeifert04 - thanks for working on this! Do you still need help with the logging stuff?

astrofrog commented:
I'm confused - isn't the first beta out?

https://www.python.org/downloads/release/python-360b1/

Does that not mean 12th September? I have 3.6b1 running on my computer :)

astrofrog commented:
By the way, you can install the beta using the ``conda-forge/label/prerelease`` channel. I think you can just revert the changes to the Travis config, then add ``conda-forge/label/prerelease`` to ``CONDA_CHANNELS`` and set a build to have PYTHON_VERSION=3.6?

MSeifert04 commented:
@astrofrog 

> Do you still need help with the logging stuff?

Unfortunatly yes, the `log_to_list` stuff failures are a mystery to me.

> I'm confused - isn't the first beta out?

Sorry, I don't know what I was looking at when I wrote https://github.com/astropy/astropy/pull/5489#issuecomment-263352157 but I meant the release candidate as scheduled for 2016-12-05 https://www.python.org/dev/peps/pep-0494/#schedule

> By the way, you can install the beta using the conda-forge/label/prerelease channel.

Last I checked they only had `b3` while travis uses `b4+` (nightly builds). Seems like conda-forge had `b4` released just yesterday.

Any reason to switch? The 2minute overhead when building cython seems not too much of a problem with the 1 one element matrix. Also I didn't get it to run properly because some dependencies were not avaiable on conda-forge for python 3.6: numpy/cython.

astrofrog commented:
> Any reason to switch? The 2minute overhead when building cython seems not too much of a problem with the 1 one element matrix. Also I didn't get it to run properly because some dependencies were not avaiable on conda-forge for python 3.6: numpy/cython.

Nope sorry, I thought maybe Travis wasn't using the beta versions yet because of your question - just ignore me :)

I'll see if I can take a look at the logging issues.

astrofrog commented:
Ok so there is either an undocumented change in Python 3.6 or a bug:

```python
import warnings

def deal_with_warning(*args, **kwargs):
    print("warning emitted")

with warnings.catch_warnings(record=True):
    warnings.showwarning = deal_with_warning
    warnings.warn("This is a warning")
```

does not show the print statement in Python 3.6 but does in 3.5.

astrofrog commented:
Here is the CPython commit that caused this to change:

```
commit c3b8ad7e641e6d9ebef1b3fecf3d81b24080fab7
Author: Victor Stinner <victor.stinner@gmail.com>
Date:   Sat Mar 19 00:47:17 2016 +0100

    Add _showwarnmsg() and _formatwarnmsg() to warnings
    
    Issue #26568: add new  _showwarnmsg() and _formatwarnmsg() functions to the
    warnings module.
    
    The C function warn_explicit() now calls warnings._showwarnmsg() with a
    warnings.WarningMessage as parameter, instead of calling warnings.showwarning()
    with multiple parameters.
    
    _showwarnmsg() calls warnings.showwarning() if warnings.showwarning() was
    replaced. Same for _formatwarnmsg(): call warnings.formatwarning() if it was
    replaced.

:040000 040000 bdf642a2a4bdff91064b7358e50cd504566c2de1 29d265d3fe03120c013fd5219efbc791ddb24742 M	Lib
:040000 040000 dce9b31eada18f462b7d6d173906a57156ceda5b 5524e53b697b18ba7f79650a379604411eb90cc9 M	Python
```

astrofrog commented:
It's not 100% clear if it's a CPython bug, but I opened an issue just in case: http://bugs.python.org/issue28835

MSeifert04 commented:
@astrofrog That certainly could be the cause of these failures. Thank you for investigating and reporting it! 👍

MSeifert04 commented:
The remaining failures are (provisionally) marked as release-blocker for python. I think it would be good to have this in before the official 3.6 is released (~2 weeks) so 1.3 would be the wrong version for this, right?

bsipocz commented:
@MSeifert04 @astrofrog @mhvk - What about getting all the fixes that work into v1.3? In that case we still have a slight chance that the 1.3 will be 3.6 compatible (given that the logger stuff gets ironed out upstream), while delaying this PR would definitely mean for certain that it isn't fully compatible.
Would it take much to clear out what's not working and the travis file?

MSeifert04 commented:
@bsipocz I reverted the Travis changes and added a changelog entry. Just in case you want to get this in.

Apart from the upstream reported issues everything worked fine on 3.6 and they probably won't change/break stuff few days before their release, right? 😅

MSeifert04 commented:
Very weird failures, I guess that's somewhat related to the astropy server being down: #5547.

MSeifert04 commented:
No new failures in 3.6 and no failures in the current tests.

I don't know if you want to wait for python RCs or merge this before the astropy RC is out. But it should be mergeable now.

mhvk commented:
It looks all OK to me, so I'll go ahead and merge (travis errors are due to a timeout and me cancelling the OS build).

mhvk commented:
Looks like the python regression (http://bugs.python.org/issue28835) got fixed -- so, we astropy should now work out of the box with `rc1`! Thanks, @MSeifert04!
